### PR TITLE
Add times to evaluation period (Feature 812) (WIP)

### LINKED ISSUE #812: Add times to evaluation period
The evaluation period could be defined by days and time instead of just days like currently implemented. That would allow to configure a precisely timed start/end for the evaluation period.

For this, the cron job checking for the start of an evaluation has to run every minute and the documentation must be updated accordingly.

### PR DESCRIPTION
**Pull request by schumannd**
Fixes  #812 based on #944

### CODE REVIEW

File: evap/grades/tests.py [hunk 1]
```diff
@@ -22,19 +22,21 @@ def setUpTestData(cls):
...
+                                vote_start_date=datetime.datetime.now() - datetime.timedelta(10),
+                                vote_end_date=datetime.datetime.now() + datetime.timedelta(10),
+                                participants=[
+                                    UserProfile.objects.get(username="student"),
+                                    UserProfile.objects.get(username="student2"),
+                                    UserProfile.objects.get(username="student3"),
+                                ],
+                                voters=[
+                                    UserProfile.objects.get(username="student"),
+                                    UserProfile.objects.get(username="student2"),
```
Comment (karyon) [inline evap/grades/tests.py]: please put the students into variables and use those here.

...your fault if you're doing random formatting changes :)

File: evap/grades/tests.py [hunk 2]
```diff
@@ -122,7 +124,8 @@ def test_upload_final_grades(self):
         # state: reviewed
         course.review_finished()
         course.save()
-        self.helper_check_final_grade_upload(course, course.num_participants + course.contributions.exclude(contributor=None).count())
+        self.helper_check_final_grade_upload(course, course.num_participants + course.contributions.exclude(
+            contributor=None).count())
```
Comment (karyon) [inline evap/grades/tests.py]: that's really not necessary

File: evap/grades/tests.py [hunk 3]
```diff
@@ -131,20 +134,21 @@ def test_upload_final_grades(self):
...
+                            vote_start_date=datetime.datetime.now(),
+                            state="reviewed",
+                            participants=[
+                                UserProfile.objects.get(username="student"),
+                                UserProfile.objects.get(username="student2"),
+                                UserProfile.objects.get(username="student3"),
+                            ],
+                            voters=[
+                                UserProfile.objects.get(username="student"),
+                                UserProfile.objects.get(username="student2"),
```
Comment (karyon) [inline evap/grades/tests.py]: same here

File: evap/grades/tests.py [hunk 4]
```diff
@@ -158,7 +162,8 @@ def test_toggle_no_grades(self):
         self.assertTrue(course.gets_no_grade_documents)
         # course should get published here
         self.assertEqual(course.state, "published")
-        self.assertEqual(len(mail.outbox), course.num_participants + course.contributions.exclude(contributor=None).count())
+        self.assertEqual(len(mail.outbox),
+                         course.num_participants + course.contributions.exclude(contributor=None).count())
```
Comment (karyon) [inline evap/grades/tests.py]: same here

File: evap/staff/forms.py [hunk 5]
```diff
@@ -234,8 +235,9 @@ def __init__(self, *args, **kwargs):
     def save(self, *args, **kw):
         user = kw.pop("user")
         self.instance.last_modified_user = user
-        self.instance.vote_start_date = self.cleaned_data['event_date']
-        self.instance.vote_end_date = self.cleaned_data['event_date']
+        the_date = self.cleaned_data['event_date']
```
Comment (karyon) [inline evap/staff/forms.py]: variable name, suggestion: `event_date`

File: evap/settings.py [hunk 6]
```diff
@@ -265,6 +265,8 @@
     ('ß', 'ss'),
 ]
 
+# Specify an offset that will be added to the evaluation end date (e.g. 3: If the end date is 01.01., the evaluation will end at 02.01. 03:00.).
+EVALUATION_END_OFFSET = 3
```
Comment (janno42) [inline evap/settings.py]: this should be sth. like `EVALUATION_END_OFFSET_HOURS`

File: evap/evaluation/templates/base.html [hunk 7]
```diff
@@ -210,8 +211,8 @@ <h4 class="modal-title">{% trans "Feedback" %}</h4>
     {% compress js %}
         <script type="text/javascript" src="{% get_static_prefix %}js/jquery-2.1.3.min.js"></script>
         <script type="text/javascript" src="{% get_static_prefix %}js/jquery-ui-1.10.3.min.js"></script>
```
Comment (karyon) [inline evap/evaluation/templates/base.html]: please remove jquery ui, it is most likely unusused now.

File: evap/evaluation/models.py [hunk 8]
```diff
@@ -414,7 +422,7 @@ def days_left_for_evaluation(self):
 
     @property
     def days_until_evaluation(self):
-        return (self.vote_start_date - datetime.date.today()).days
+        return (self.vote_start_date - datetime.datetime.now()).days
```
Comment (karyon) [inline evap/evaluation/models.py]: as per @janno42's request, can you change this to only take days into account, not the time?

File: evap/evaluation/models.py [hunk 9]
```diff
@@ -525,19 +533,18 @@ def grades_activated(self):
...
         from evap.evaluation.tools import send_publish_notifications
-        today = datetime.date.today()
 
         courses_new_in_evaluation = []
         evaluation_results_courses = []
 
         for course in cls.objects.all():
             try:
-                if course.state == "approved" and course.vote_start_date <= today:
+                if course.state == "approved" and course.vote_start_date.date() <= datetime.date.today():
```
Comment (karyon) [inline evap/evaluation/models.py]: now that's the one place where we actually need to use the time and not just the date, isn't it?

File: evap/evaluation/models.py [hunk 10]
```diff
@@ -525,19 +533,18 @@ def grades_activated(self):
...
                     course.evaluation_begin()
                     course.last_modified_user = UserProfile.cronjob_user()
                     course.save()
                     courses_new_in_evaluation.append(course)
-                elif course.state == "in_evaluation" and course.vote_end_date < today:
+                elif course.state == "in_evaluation" and course.actual_evaluation_end_datetime < datetime.datetime.now():
```
Comment (karyon) [inline evap/evaluation/models.py]: can you turn the expression around? (with >=)

File: evap/evaluation/tests/test_commands.py [hunk 11]
```diff
@@ -85,14 +85,15 @@ def test_dumpdata_called(self):
 
 @override_settings(REMIND_X_DAYS_AHEAD_OF_END_DATE=[0, 2])
 class TestSendRemindersCommand(TestCase):
+    now = datetime.datetime.now()
```
Comment (karyon) [inline evap/evaluation/tests/test_commands.py]: can you decide on one import style (`import datetime` or `from datetime import datetime`) and apply that consistently everywhere? (i think i prefer the second style in this case because otherwise the code gets a lot of redundency)

File: evap/grades/tests.py [hunk 12]
```diff
@@ -158,7 +150,8 @@ def test_toggle_no_grades(self):
         self.assertTrue(course.gets_no_grade_documents)
         # course should get published here
         self.assertEqual(course.state, "published")
-        self.assertEqual(len(mail.outbox), course.num_participants + course.contributions.exclude(contributor=None).count())
+        self.assertEqual(
+                len(mail.outbox), course.num_participants + course.contributions.exclude(contributor=None).count())
```
Comment (karyon) [inline evap/grades/tests.py]: plz no

File: evap/evaluation/models.py [hunk 13]
```diff
@@ -226,7 +228,7 @@ class Course(models.Model, metaclass=LocalizeModelBase):
     _voter_count = models.IntegerField(verbose_name=_("voter count"), blank=True, null=True, default=None)
 
     # when the evaluation takes place
-    vote_start_date = models.DateField(verbose_name=_("first day of evaluation"))
+    vote_start_date = models.DateTimeField(verbose_name=_("start of evaluation"))
```
Comment (karyon) [inline evap/evaluation/models.py]: i would suggest renaming this to vote_start_datetime, what do you think?

File: evap/evaluation/models.py [hunk 14]
```diff
@@ -266,9 +268,15 @@ def is_not_fully_reviewed(self):
         return self.open_textanswer_set.exists()
 
     @property
+    def actual_evaluation_end_datetime(self):
```
Comment (karyon) [inline evap/evaluation/models.py]: i would want this to be called actual_vote_end_datetime to be consistent with vote_*_date, but what if we just called this vote_end_datetime?

File: evap/staff/forms.py [hunk 15]
```diff
@@ -178,7 +179,7 @@ def clean(self):
         vote_start_date = self.cleaned_data.get('vote_start_date')
         vote_end_date = self.cleaned_data.get('vote_end_date')
         if vote_start_date and vote_end_date:
-            if vote_start_date >= vote_end_date:
+            if vote_start_date.date() >= vote_end_date:
```
Comment (karyon) [inline evap/staff/forms.py]: this was the point where we discussed whether to allow evaluations starting and ending on the same day, right? was there any reason to not allow that? if you change this (`>=` to `>`), please do so in a separate commit.

File: evap/staff/forms.py [hunk 16]
```diff
@@ -230,8 +231,9 @@ def __init__(self, *args, **kwargs):
     def save(self, *args, **kw):
         user = kw.pop("user")
         self.instance.last_modified_user = user
-        self.instance.vote_start_date = self.cleaned_data['event_date']
-        self.instance.vote_end_date = self.cleaned_data['event_date']
+        event_date = self.cleaned_data['event_date']
+        self.instance.vote_start_date = datetime.datetime(event_date.year, event_date.month, event_date.day)
```
Comment (karyon) [inline evap/staff/forms.py]: you can use `date_to_datetime` for that right?

File: evap/staff/forms.py (hunk 15 above)
Comment (karyon) [inline evap/staff/forms.py]: hah no, i just saw you changed this in contributor/forms.py. this here is the other place where this needs to be applied :)

File: evap/evaluation/models.py (hunk 9 above)
Comment (schumannd) [inline evap/evaluation/models.py]: no, the start date is a date. only the end date is a datetime

Comment (karyon) [inline evap/evaluation/models.py]: ...it's exactly the other way round as far as i can see.

Comment (schumannd) [inline evap/evaluation/models.py]: you are right

File: evap/evaluation/management/commands/send_reminders.py [hunk 17]
```diff
@@ -29,7 +29,7 @@ def handle(self, *args, **options):
 
         for recipient in recipients:
             due_courses = dict()
-            for course in Course.objects.filter(participants=recipient, state='in_evaluation').exclude(voters=recipient):
+            for course in Course.objects.filter(participants=recipient, state='in_evaluation', vote_end_date__in=check_dates).exclude(voters=recipient):
```
Comment (schumannd) [inline evap/evaluation/management/commands/send_reminders.py]: I guess students simply got reminded to vote on all their unvoted courses, as long as one of them was ending in REMIND_X_DAYS_AHEAD_OF_END_DATE.  Do we want to keep that behavior?

Comment (janno42) [inline evap/evaluation/management/commands/send_reminders.py]: yes, the reminder email is triggered if one course fulfills the condition but must list all courses that are still due.

File: evap/evaluation/migrations/0057_auto_20170619_2238.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
```
Comment (karyon) [inline evap/evaluation/migrations/0057_auto_20170619_2238.py]: can you give the migration file a helpful name?

File: evap/contributor/forms.py [hunk 19]
```diff
@@ -57,8 +57,7 @@ def clean_vote_end_date(self):
 
         # The actual deadline is EVALUATION_END_OFFSET_HOURS:00 AM of the day after vote_end_date.
         # Therefore an evaluation date 24h + EVALUATION_END_OFFSET_HOURS in the past would technically still be in the future.
-        if vote_end_date and date_to_datetime(vote_end_date) < \
-                        datetime.now() - timedelta(hours=24 + EVALUATION_END_OFFSET_HOURS):
+        if vote_end_date and date_to_datetime(vote_end_date) - timedelta(hours=24 + EVALUATION_END_OFFSET_HOURS) < datetime.now():
```
Comment (karyon) [inline evap/contributor/forms.py]: i think that must be a +, not -

File: evap/contributor/forms.py [hunk 20]
```diff
@@ -46,13 +47,16 @@ def clean(self):
...
+        if vote_start_date and vote_start_date < datetime.datetime.now():
             raise forms.ValidationError(_("The first day of evaluation must be in the future."))
         return vote_start_date
 
     def clean_vote_end_date(self):
-        vote_end_date = self.cleaned_data.get('vote_end_date')
-        if vote_end_date and vote_end_date < datetime.date.today():
+        vote_end_date_unchanged = self.cleaned_data.get('vote_end_date')
+        vote_end_date_unchanged.replace(hour=23, minute=59)
+        vote_end_date = vote_end_date_unchanged.replace(hour=23, minute=59) + \
```
Comment (schumannd) [inline evap/contributor/forms.py]: it is necessary

File: evap/evaluation/migrations/0057_auto_20170619_2238.py [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    operations = [
+        migrations.RemoveField(
+            model_name='course',
+            name='vote_start_date',
+        ),
+        migrations.AddField(
+            model_name='course',
+            name='vote_start_datetime',
+            field=models.DateTimeField(default=timezone.now(), verbose_name='start of evaluation'),
+            preserve_default=False,
```
Comment (karyon) [inline evap/evaluation/migrations/0057_auto_20170619_2238.py]: this will discard all the data as far as i can see. i think you actually need to have an AlterField and RenameField.

File: evap/evaluation/templates/base.html [hunk 22]
```diff
@@ -209,9 +210,8 @@ <h4 class="modal-title">{% trans "Feedback" %}</h4>
 
     {% compress js %}
         <script type="text/javascript" src="{% get_static_prefix %}js/jquery-2.1.3.min.js"></script>
-        <script type="text/javascript" src="{% get_static_prefix %}js/jquery-ui-1.10.3.min.js"></script>
```
Comment (karyon) [inline evap/evaluation/templates/base.html]: please remove the actual file as well.

File: evap/evaluation/models.py [hunk 23]
```diff
@@ -410,11 +418,11 @@ def responsible_contributors(self):
 
     @property
     def days_left_for_evaluation(self):
-        return (self.vote_end_date - datetime.date.today()).days
+        return (self.vote_end_date - date.today()).days
```
Comment (karyon) [inline evap/evaluation/models.py]: as per @janno42's request, can you change this to only take days into account, not the time?

File: evap/evaluation/models.py [hunk 24]
```diff
@@ -410,11 +418,11 @@ def responsible_contributors(self):
...
 
     @property
     def days_until_evaluation(self):
-        return (self.vote_start_date - datetime.date.today()).days
+        return (self.vote_start_datetime - date.today()).days
```
Comment (karyon) [inline evap/evaluation/models.py]: same here

File: evap/evaluation/tests/test_commands.py [hunk 25]
```diff
@@ -85,15 +85,16 @@ def test_dumpdata_called(self):
 
 @override_settings(REMIND_X_DAYS_AHEAD_OF_END_DATE=[0, 2])
 class TestSendRemindersCommand(TestCase):
-    today = datetime.date.today()
+    now = datetime.now()
+    today = date.today()
```
Comment (karyon) [inline evap/evaluation/tests/test_commands.py]: lets remove they, they don't provide much value anymore.

File: evap/staff/forms.py [hunk 26]
```diff
@@ -1,3 +1,4 @@
+import datetime
```
Comment (karyon) [inline evap/staff/forms.py]: unused import

File: evap/grades/tests.py [hunk 27]
```diff
@@ -158,7 +150,7 @@ def test_toggle_no_grades(self):
         self.assertTrue(course.gets_no_grade_documents)
         # course should get published here
         self.assertEqual(course.state, "published")
-        self.assertEqual(len(mail.outbox), course.num_participants + course.contributions.exclude(contributor=None).count())
+        self.assertEqual(len(mail.outbox),course.num_participants + course.contributions.exclude(contributor=None).count())
```
Comment (karyon) [inline evap/grades/tests.py]: missing space

**Comment by karyon** [comment]
* please use moment.min.js without the locales and add the german locale. the file size difference is massive (17kb vs 66kb)
* please remove jquery.ui.css as well
* the comment "as per @janno42's request, can you change this to only take days into account, not the time?" and the one below that are still valid
* travis still fails

**Comment by karyon** [comment]
on @schumannd's request i took over and fixed the remaining stuff. @janno42 please have another look.

File: evap/grades/tests.py [hunk 28]
```diff
@@ -56,11 +52,11 @@ def helper_upload_grades(self, course, final_grades):
 
         final = "?final=true" if final_grades else ""
         response = self.app.post(
-            "/grades/semester/{}/course/{}/upload{}".format(course.semester.id, course.id, final),
-            params={"description_en": "Grades", "description_de": "Grades"},
-            user="grade_publisher",
-            content_type='multipart/form-data',
-            upload_files=upload_files,
+                "/grades/semester/{}/course/{}/upload{}".format(course.semester.id, course.id, final),
```
Comment (janno42) [inline evap/grades/tests.py]: double indentation

File: evap/grades/tests.py [hunk 29]
```diff
@@ -122,29 +118,25 @@ def test_upload_final_grades(self):
...
-                UserProfile.objects.get(username="student"),
-                UserProfile.objects.get(username="student2"),
-                UserProfile.objects.get(username="student3"),
-            ],
-            voters=[
-                UserProfile.objects.get(username="student"),
-                UserProfile.objects.get(username="student2"),
-            ]
+        course = mommy.make(
+                Course,
```
Comment (janno42) [inline evap/grades/tests.py]: double indentation

File: evap/grades/tests.py [hunk 30]
```diff
@@ -16,25 +16,21 @@ class GradeUploadTests(WebTest):
...
-                UserProfile.objects.get(username="student"),
-                UserProfile.objects.get(username="student2"),
-                UserProfile.objects.get(username="student3"),
-            ],
-            voters=[
-                UserProfile.objects.get(username="student"),
-                UserProfile.objects.get(username="student2"),
-            ]
+        cls.course = mommy.make(
+                Course,
```
Comment (janno42) [inline evap/grades/tests.py]: double indentation

File: evap/evaluation/templates/base.html [hunk 31]
```diff
@@ -274,8 +275,17 @@ <h4 class="modal-title">{% trans "Feedback" %}</h4>
         };
         apply_select2_disabled($('select[disabled]').not('.form-template select'));
 
-        // run the jquery datepicker plugin
-        $("input[name*='date']:not([readonly='True'])").datepicker( $.datepicker.regional["{{ LANGUAGE_CODE }}"] );
+        // run the bootstrap datepicker plugin
+        $("input[name*='end_date']:not([readonly='True'])").datetimepicker({
+                locale: "{{ LANGUAGE_CODE }}",
```
Comment (janno42) [inline evap/evaluation/templates/base.html]: double indentation

File: evap/evaluation/templates/base.html [hunk 32]
```diff
@@ -274,8 +275,17 @@ <h4 class="modal-title">{% trans "Feedback" %}</h4>
...
+                format: "YYYY-MM-DD",
+                sideBySide: true
+            });
+        $("input[name*='start_date']:not([readonly='True'])").datetimepicker({
+                locale: "{{ LANGUAGE_CODE }}",
```
Comment (janno42) [inline evap/evaluation/templates/base.html]: double indentation

File: evap/evaluation/tests/test_commands.py [hunk 33]
```diff
@@ -85,15 +85,13 @@ def test_dumpdata_called(self):
 
 @override_settings(REMIND_X_DAYS_AHEAD_OF_END_DATE=[0, 2])
 class TestSendRemindersCommand(TestCase):
-    today = datetime.date.today()
-
     def test_remind_user_about_one_course(self):
         user_to_remind = mommy.make(UserProfile)
         course = mommy.make(
                 Course,
```
Comment (janno42) [inline evap/evaluation/tests/test_commands.py]: double indentation

File: evap/evaluation/tests/test_commands.py [hunk 34]
```diff
@@ -107,14 +105,14 @@ def test_remind_user_once_about_two_courses(self):
         course1 = mommy.make(
                 Course,
```
Comment (janno42) [inline evap/evaluation/tests/test_commands.py]: double indentation

File: evap/evaluation/tests/test_commands.py [hunk 35]
```diff
@@ -107,14 +105,14 @@ def test_remind_user_once_about_two_courses(self):
...
                 state='in_evaluation',
-                vote_start_date=self.today - datetime.timedelta(days=1),
-                vote_end_date=self.today + datetime.timedelta(days=0),
+                vote_start_datetime=datetime.now() - timedelta(days=1),
+                vote_end_date=date.today() + timedelta(days=0),
                 participants=[user_to_remind])
         course2 = mommy.make(
                 Course,
```
Comment (janno42) [inline evap/evaluation/tests/test_commands.py]: double indentation

File: evap/evaluation/tests/test_commands.py [hunk 36]
```diff
@@ -128,8 +126,8 @@ def test_dont_remind_already_voted(self):
         mommy.make(
                 Course,
```
Comment (janno42) [inline evap/evaluation/tests/test_commands.py]: double indentation
